cmd/jujud/agent/safemode.go

```diff
@@ -0,0 +1,390 @@
+// Copyright 2012, 2013 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package agent
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/juju/clock"
+	"github.com/juju/cmd/v3"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"github.com/juju/loggo"
+	"github.com/juju/names/v4"
+	"github.com/juju/utils/v3/voyeur"
+	"github.com/juju/version/v2"
+	"github.com/juju/worker/v3"
+	"github.com/juju/worker/v3/dependency"
+
+	"github.com/juju/juju/agent"
+	agentconfig "github.com/juju/juju/agent/config"
+	agentengine "github.com/juju/juju/agent/engine"
+	agenterrors "github.com/juju/juju/agent/errors"
+	k8sconstants "github.com/juju/juju/caas/kubernetes/provider/constants"
+	jujucmd "github.com/juju/juju/cmd"
+	"github.com/juju/juju/cmd/internal/agent/agentconf"
+	"github.com/juju/juju/cmd/jujud/agent/safemode"
+	"github.com/juju/juju/cmd/jujud/reboot"
+	cmdutil "github.com/juju/juju/cmd/jujud/util"
+	"github.com/juju/juju/internal/storage/looputil"
+	"github.com/juju/juju/rpc/params"
+	jworker "github.com/juju/juju/worker"
+	"github.com/juju/juju/worker/dbaccessor"
+	"github.com/juju/juju/worker/logsender"
+)
+
+// NewSafeModeAgentCommand creates a Command which handles parsing
+// command-line arguments and instantiating and running a
+// MachineAgent.
+func NewSafeModeAgentCommand(
+	ctx *cmd.Context,
+	safeModeMachineAgentFactory safeModeMachineAgentFactoryFnType,
+	agentInitializer AgentInitializer,
+	configFetcher agentconfig.AgentConfigWriter,
+) cmd.Command {
+	return &safeModeAgentCommand{
+		ctx:                         ctx,
+		safeModeMachineAgentFactory: safeModeMachineAgentFactory,
+		agentInitializer:            agentInitializer,
+		currentConfig:               configFetcher,
+	}
+}
+
+type safeModeAgentCommand struct {
+	cmd.CommandBase
+
+	// This group of arguments is required.
+	agentInitializer            AgentInitializer
+	currentConfig               agentconfig.AgentConfigWriter
+	safeModeMachineAgentFactory safeModeMachineAgentFactoryFnType
+	ctx                         *cmd.Context
+
+	isCaas   bool
+	agentTag names.Tag
+
+	// The following are set via command-line flags.
+	machineId string
+	// TODO(controlleragent) - this will be in a new controller agent command
+	controllerId string
+
+	dangerous bool
+}
+
+// Init is called by the cmd system to initialize the structure for
+// running.
+func (a *safeModeAgentCommand) Init(args []string) error {
+	if !a.dangerous {
```

Comment:
Let's drop this. If the user is hand-running in this specific mode, we can ensure they acknowledge the out-of-band nature.